kore/src/Kore/Step/Implication.hs

```diff
@@ -0,0 +1,440 @@
+            } = implication'
+
+instance TopBottom (Implication modality) where
+    isTop _ = False
+    isBottom _ = False
+
+instance From (Implication modality) Attribute.PriorityAttributes where
+    from = from @(Attribute.Axiom _ _) . attributes
+
+instance From (Implication modality) Attribute.HeatCool where
+    from = from @(Attribute.Axiom _ _) . attributes
+
+freeVariablesRight
+    :: Implication modality
+    -> FreeVariables RewritingVariableName
+freeVariablesRight implication' =
+    freeVariables
+        ( TermLike.mkExistsN existentials
+            (OrPattern.toTermLike right)
+        )
+  where
+    Implication { right, existentials } = implication'
+
+freeVariablesLeft
+    :: Implication modality
+    -> FreeVariables RewritingVariableName
+freeVariablesLeft implication' =
+    freeVariables left
+  where
+    Implication { left } = implication'
+
+instance HasFreeVariables (Implication modality) RewritingVariableName where
+    freeVariables implication' =
+        freeVariablesLeft implication'
+        <> freeVariablesRight implication'
+
+-- | Creates an 'Implication' from a left hand side 'Pattern'
+-- and an 'OrPattern', representing the right hand side pattern.
+-- The list of element variables are existentially quantified
+-- in the right hand side.
+implication
+    :: modality
+    -> Pattern RewritingVariableName
+    -> OrPattern RewritingVariableName
+    -> [ElementVariable RewritingVariableName]
+    -> Implication modality
+implication modality left right existentials =
+    Implication
+        { left
+        , right
+        , modality
+        , existentials
+        , attributes = Default.def
+        }
+
+{- | Construct a 'TermLike' from the parts of an implication-based rule.
+
+The 'TermLike' has the following form:
+
+@
+\\implies{S}(\\and{S}(left, requires), alias{S}(right))
+@
+
+that is,
+
+@
+left ∧ requires → alias(right)
+@
+
+ -}
+implicationToTerm
+    :: Modality
+    -> Implication modality
+    -> TermLike VariableName
+implicationToTerm modality representation =
+    TermLike.mkImplies
+        (TermLike.mkAnd leftCondition leftTerm)
+        (TermLike.applyModality modality rightPattern)
+  where
+    Implication { left, right, existentials } = representation
+    leftTerm =
+        Pattern.term left
+        & getRewritingTerm
+    sort = TermLike.termLikeSort leftTerm
+    leftCondition =
+        Pattern.withoutTerm left
+        & Pattern.fromCondition sort
+        & Pattern.toTermLike
+        & getRewritingTerm
+    rightPattern =
+        TermLike.mkExistsN existentials (OrPattern.toTermLike right)
+        & getRewritingTerm
+
+substituteRight
+    :: Map
+        (SomeVariableName RewritingVariableName)
+        (TermLike RewritingVariableName)
+    -> Implication modality
+    -> Implication modality
+substituteRight rename implication'@Implication{ right, existentials } =
+    implication'
+        { right = OrPattern.substitute subst right
+        }
```

Comment:
This implementation would capture variables appearing on the right-hand side of a substitution. Please add a unit test for this, and fix it.

Reply:
I don't see why you would think that. I did add a unit test and the function was behaving as expected...